### PR TITLE
Add ECSql tests for ECDb

### PR DESCRIPTION
**Pull request by rschili**
imodel-native: https://github.com/iTwin/imodel-native/pull/937

Framework that makes it easier to compose a large amount of SQL tests which verify results (not just prepare and done).

Part of a shared effort to increase our ECDb test coverage after recent breaks.

Unbenannt

The picture separates our API into 9 areas which we can distribute among the team.

- [x] 1 - Subquery, ClassName, MemberFunctionCall, TableValuesFunction(Naron)
- [x] 2 - CommonTable Expressions (Soham)
- [x] 3 - Properties, LimitOffset, Group by, Having, Where, SubqueryTest, AllOrAny (Rohit)
- [x] 4- Joins(Naron)
- [x] 5- Pragma and NavValueCreation (Rohit)
- [x] 6 - Parameter, Boolean, Range (Akshay) (Under review: Rohit)
- [x] 7- Literals, EnumValue, ExtractValue( Soham)
- [x]  8 - Cast, PropertyNameList, SubqueryValue, IIF (Robert)
- [x] 9 - Window Functions (Robert)

### CODE REVIEW

**Comment by rschili** [comment]
I would like to roll this out this week, but I also want to roll out the linked native PR with it.
I'm getting several failing tests with those native changes. Still, the change is one we really want to make.
It affects metadata for binary columns returned by ConcurrentQuery. It returned wrong types, but apparently some code relies on the wrong types already... Trying to track everything down and fix it.

**Comment by khanaffan** [comment]
One general comment I have is how files are organize and named. These test are for ECSQL and folders and files should be name as such. ECDb word should generally be avoided as its not something that is well known on itwin.js side. The term we use is ECSql, IModels etc.

```
ECDb -> ECSql 
.ecdb.md -> .ecsql.md 
```

Markdown test is just format what these test stand for is regression & coverage.. 

```
ECDbMarkdownDataset.ts -> ECSqlDataset.ts
...
```

Directory structure can be as following. Its like package.

```
tests/ecsql/dataset/  (binary/bim/xml)
tests/ecsql/queries/  (only .ecsql.md)
tests/ecsql/src/        (source .ts)
tests/ecsql/README.md  (root)
```

File: core/backend/src/test/assets/ECDbTests/Binders.ecdbtest.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
+Copyright © Bentley Systems, Incorporated. All rights reserved. See [LICENSE.md](../../../../LICENSE.md) for license terms and full copyright notice.
+
+# Testing integer binder
+
+- dataset: AllProperties.bim
+
+```sql
+SELECT e.i FROM aps.TestElement e where e.i > ? and e.i < ? order by e.i
```
Comment (khanaffan) [inline core/backend/src/test/assets/ECDbTests/Binders.ecdbtest.md:8]: I was expecting BETWEEN keyword `WHERE  x BETWEEN 1 AND 2`.

**Review by khanaffan — APPROVED** [review]

File: common/changes/@itwin/core-backend/rob-ecsql-tests_2024-12-13-11-28.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@itwin/core-backend",
+      "comment": "Add ECSql test suite",
```
Comment (pmconne) [inline common/changes/@itwin/core-backend/rob-ecsql-tests_2024-12-13-11-28.json:5]:
```suggestion
      "comment": "",
```

End users don't care about tests.

**Review by pmconne — APPROVED** [review]

**Comment by soham-bentley** [comment]
> One general comment I have is how files are organize and named. These test are for ECSQL and folders and files should be name as such. ECDb word should generally be avoided as its not something that is well known on itwin.js side. The term we use is ECSql, IModels etc.
> 
> ```
> ECDb -> ECSql 
> .ecdb.md -> .ecsql.md 
> ```
> 
> Markdown test is just format what these test stand for is regression & coverage..
> 
> ```
> ECDbMarkdownDataset.ts -> ECSqlDataset.ts
> ...
> ```
> 
> Directory structure can be as following. Its like package.
> 
> ```
> tests/ecsql/dataset/  (binary/bim/xml)
> tests/ecsql/queries/  (only .ecsql.md)
> tests/ecsql/src/        (source .ts)
> tests/ecsql/README.md  (root)
> ```

Test Structure updated

**Review by khanaffan — APPROVED** [review]

**Review by aruniverse — APPROVED** [review]
approving addon changes

File: core/backend/package.json [hunk 3]
```diff
@@ -93,10 +94,12 @@
     "sinon": "^17.0.2",
     "source-map-loader": "^4.0.0",
     "typescript": "~5.6.2",
+    "marked": "^14.1.3",
```
Comment (aruniverse) [inline core/backend/package.json:97]: any reason we are not using v15?

Comment (rschili) [reply core/backend/package.json:97]: No reason. 15 was released a week after I created the prototype for these tests, and we never updated that's all.

**Comment by soham-bentley** [comment]
@Mergifyio backport release/4.11.x

**Pull request merged**